sw/device/tests/edn_boot_mode.c

```diff
@@ -0,0 +1,165 @@
+  // Enable ENTROPY_SRC in Non-FIPS/FIPS mode based on the value of round.
+  entropy_src_config.fips_enable = (round != 1);
+  CHECK_DIF_OK(dif_entropy_src_configure(&entropy_src, entropy_src_config,
+                                         kDifToggleEnabled));
+  // Enable CSRNG.
+  CHECK_DIF_OK(dif_csrng_configure(&csrng));
+
+  if (round == 1) {
+    // Enable EDN1 (the one connected to OTBN RND) in boot-time request mode.
+    CHECK_DIF_OK(dif_edn_set_boot_mode(&edn1));
+    CHECK_DIF_OK(dif_edn_configure(&edn1));
+    CHECK_STATUS_OK(
+        edn_testutils_wait_for_state(&edn1, kDifEdnSmStateBootGenAckWait));
+    // Re-enable ENTROPY_SRC in FIPS mode.
+    CHECK_DIF_OK(dif_entropy_src_stop(&entropy_src));
+    entropy_src_config.fips_enable = true;
+    CHECK_DIF_OK(dif_entropy_src_configure(&entropy_src, entropy_src_config,
+                                           kDifToggleEnabled));
+    // Enable EDN0 in auto request mode.
+    CHECK_DIF_OK(dif_edn_set_auto_mode(&edn0, edn_params0));
+  }
+
+  if (round == 2) {
+    // Enable EDN1 (the one connected to OTBN RND) in auto request mode.
+    CHECK_DIF_OK(dif_edn_set_auto_mode(&edn1, edn_params1));
+    CHECK_STATUS_OK(
+        edn_testutils_wait_for_state(&edn1, kDifEdnSmStateAutoAckWait));
+    // Re-enable ENTROPY_SRC in Non-FIPS mode.
+    CHECK_DIF_OK(dif_entropy_src_stop(&entropy_src));
+    entropy_src_config.fips_enable = false;
+    CHECK_DIF_OK(dif_entropy_src_configure(&entropy_src, entropy_src_config,
+                                           kDifToggleEnabled));
+    // Enable EDN0 in boot-time request mode.
+    CHECK_DIF_OK(dif_edn_set_boot_mode(&edn0));
+    CHECK_DIF_OK(dif_edn_configure(&edn0));
+    CHECK_STATUS_OK(
+        edn_testutils_wait_for_state(&edn0, kDifEdnSmStateBootGenAckWait));
+  }
+
+  if (round == 3) {
+    // Enable both EDNs in auto request mode.
+    CHECK_DIF_OK(dif_edn_set_auto_mode(&edn0, edn_params0));
+    CHECK_DIF_OK(dif_edn_set_auto_mode(&edn1, edn_params1));
+  }
+}
+
+static void consume_entropy(unsigned int round,
+                            dif_otbn_err_bits_t otbn_err_val,
+                            dif_rv_core_ibex_rnd_status_t ibex_rnd_fips) {
+  uint32_t ibex_rnd_data;
+  dif_rv_core_ibex_rnd_status_t ibex_rnd_status;
+  dif_otbn_irq_state_snapshot_t intr_state;
+  entropy_config(round);
+  // Launch an OTBN program consuming entropy via both
+  // the RND and the URND interface.
+  CHECK_STATUS_OK(otbn_testutils_execute(&otbn));
+  // Verify that the OTBN finishes with the expected error values
+  // and interrupt flags.
+  CHECK_STATUS_OK(otbn_testutils_wait_for_done(&otbn, otbn_err_val));
+  CHECK_DIF_OK(dif_otbn_irq_get_state(&otbn, &intr_state));
+  CHECK(intr_state & 0x1);
+  CHECK_DIF_OK(dif_otbn_irq_acknowledge_all(&otbn));
+  // Read rnd data through the IBEX and verify if the FIPS compliance
+  // status is as expected.
+  CHECK_DIF_OK(dif_rv_core_ibex_read_rnd_data(&rv_core_ibex, &ibex_rnd_data));
```

Comment:
Here it's probably worth to add a comment saying that the first read to rnd_data is to actually remove any data that could be there from a previous execution.

Reply:
Good point! I'll add the comment ASAP.